src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/PropagateCancellationTokensWhenPossible.cs

```diff
@@ -33,11 +39,151 @@ public abstract class PropagateCancellationTokensWhenPossibleAnalyzer : Diagnost
+            else
+            {
+                // The target method accepts a CancellationToken. Check if default(CancellationToken) or CancellationToken.None is being passed.
+                var argumentValue = invocation.Arguments
+                    .Select(argument => argument.Value)
+                    .FirstOrDefault(value => value.Type == cancellationTokenType);
+
+                // All valid code should result in 'argumentValue' being non-null. Even if the CancellationToken is an optional
+                // parameter and the caller does not explicitly pass it, 'argumentValue' will still correspond to a compiler-generated
+                // CancellationToken value. However, check for null to ensure we don't crash on invalid code.
+                if (argumentValue == null)
+                {
+                    // The code is invalid. The compiler error will be enough warning for the developer.
+                    return false;
+                }
+
+                if (!IsCancellationTokenNone(argumentValue, cancellationTokenType))
+                {
+                    // A non-default CancellationToken is being passed.
+                    return false;
+                }
+            }
+
+            // Check that a suitable variable of type CancellationToken is in scope.
+            if (!IsCancellationTokenVariableInScope(invocation, cancellationTokenType, compilation))
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        private static bool IsCancellationTokenNone(IOperation value, INamedTypeSymbol cancellationTokenType)
+        {
+            // This method returns true for CancellationToken.None, explicitly-passed default(CancellationToken),
+            // and compiler-generated default(CancellationToken) (e.g. as the value of an optional parameter).
+            // Note that default(CancellationToken) is equivalent to CancellationToken.None.
+
+            Debug.Assert(value.Type == cancellationTokenType);
+
+            switch (value.Kind)
+            {
+                case OperationKind.DefaultValue:
+                    return true;
+                case OperationKind.PropertyReference:
+                    var property = ((IPropertyReferenceOperation)value).Property;
+                    if (property.Name == "None" && property.ContainingType == cancellationTokenType)
+                    {
+                        return true;
+                    }
+                    break;
+            }
+
+            return false;
+        }
+
+        private static bool IsCancellationTokenVariableInScope(IInvocationOperation invocation, INamedTypeSymbol cancellationTokenType, Compilation compilation)
+        {
+            var semanticModel = compilation.GetSemanticModel(invocation.Syntax.SyntaxTree);
+            int position = invocation.Syntax.SpanStart;
+            return semanticModel.LookupSymbols(position).Any(IsUsableCancellationTokenVariable);
```

Comment:
Nit: because this is only used once, I feel like this read better as a lambda, rather than a local function.

Reply:
I prefer separating things into another function when it can't be written as a one-liner (i.e. needs an extra level of braces).